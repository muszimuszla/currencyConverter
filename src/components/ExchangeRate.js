import React, { Component } from "react";
import { connect } from "react-redux";
import { changeExchangeRate } from "../redux/actions";

export class ExchangeRate extends Component {
  handleChange = (e) => {
    if (e.target.value >= 0) {
      this.props.changeExchangeRate(e.target.value);
    } else {
      e.target.value = "";
    }
  };

  render() {
    return (
      <React.Fragment>
        <div className="exchange-rate">
          <p>1 EUR = </p>
          <input
            type="number"
            value={this.props.rate}
            name="rate"
            onChange={this.handleChange}
            min="0"
            step="0.01"
          />
          <p>PLN </p>
        </div>
      </React.Fragment>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    rate: state.rate,
  };
};

const mapDispatchToProprs = {
  changeExchangeRate: changeExchangeRate,
};

export default connect(mapStateToProps, mapDispatchToProprs)(ExchangeRate);
