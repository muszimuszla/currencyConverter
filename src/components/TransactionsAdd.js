import React, { Component } from "react";
import { connect } from "react-redux";
import { addTransaction, addToSum } from "../redux/actions";

export class TransactionsAdd extends Component {
  handleChange = (e) => {
    this.setState({ [e.target.name]: e.target.value });
  };

  handleSubmission = (e) => {
    e.preventDefault();

    let { title, euro } = this.state;
    this.props.addTransaction(title, euro);
    this.props.addToSum(euro * 1);

    document.getElementById("form").reset();
  };

  render() {
    return (
      <React.Fragment>
        <form id="form" onSubmit={this.handleSubmission}>
          <div className="add-form">
            <input
              id="title"
              placeholder="TRANSACTION TITLE"
              type="text"
              name="title"
              required
              onChange={this.handleChange}
            />
            <input
              id="EUR"
              placeholder="EUR"
              type="number"
              min="0"
              step="0.01"
              name="euro"
              required
              onChange={this.handleChange}
            />
            <button type="submit">
              <p>ADD</p>
            </button>
          </div>
        </form>
      </React.Fragment>
    );
  }
}

export default connect(null, {
  addTransaction: addTransaction,
  addToSum: addToSum,
})(TransactionsAdd);
