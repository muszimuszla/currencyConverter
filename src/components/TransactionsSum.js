import React, { Component } from "react";
import { connect } from "react-redux";

export class TransactionsSum extends Component {
  render() {
    return (
      <React.Fragment>
        <div className="sum">
          <p id="EUR">{this.props.sum.toFixed(2)}€</p>
          <p id="PLN">{(this.props.sum * this.props.rate).toFixed(2)}zł</p>
        </div>
      </React.Fragment>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    rate: state.rate,
    sum: state.sum,
  };
};

export default connect(mapStateToProps, null)(TransactionsSum);
