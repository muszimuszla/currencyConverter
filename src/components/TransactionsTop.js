import React, { Component } from "react";
import { connect } from "react-redux";

export class TransactionsTop extends Component {
  render() {
    function calculatePLN(euro, rate) {
      return (euro * rate).toFixed(2);
    }

    function calculateEURO(euro) {
      return (euro * 1).toFixed(2);
    }

    var topTransaction = { title: "", euro: 0 };
    this.props.transactions.forEach((transaction) => {
      if (parseInt(transaction.euro, 10) > topTransaction.euro) {
        topTransaction.title = transaction.title;
        topTransaction.euro = transaction.euro;
      }
    });

    return (
      <React.Fragment>
        <div className="top-transaction">
          <p id="top">TOP</p>
          <p> {topTransaction.title} </p>
          <div className="top-transaction__value">
            <p> {calculateEURO(topTransaction.euro)} EUR </p>
            <p>{calculatePLN(topTransaction.euro, this.props.rate)} PLN</p>{" "}
          </div>
        </div>
      </React.Fragment>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    transactions: state.transactions,
    rate: state.rate,
  };
};

export default connect(mapStateToProps, null)(TransactionsTop);
