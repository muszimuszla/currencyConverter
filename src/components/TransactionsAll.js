import React, { Component } from "react";
import { connect } from "react-redux";
import { removeTransaction } from "../redux/actions";
import TransactionsAdd from "./TransactionsAdd";
import TransactionsSum from "./TransactionsSum";

export class TransactionsAll extends Component {
  render() {
    function calculatePLN(euro, rate) {
      return (euro * rate).toFixed(2);
    }

    function calculateEURO(euro) {
      return (euro * 1).toFixed(2);
    }

    const transactionsItems = this.props.transactions.map(
      (transaction, index) => (
        <div key={index} className="transactions__item">
          <p id="title">{transaction.title}</p>
          <p id="EUR">{calculateEURO(transaction.euro)}€</p>
          <p id="PLN">{calculatePLN(transaction.euro, this.props.rate)}zł</p>
          <button
            className="delete"
            onClick={() =>
              this.props.removeTransaction(index, transaction.euro)
            }
          >
            X
          </button>
        </div>
      )
    );

    return (
      <React.Fragment>
        <div className="transactions">
          <TransactionsAdd />
          <div className="transactions__items">{transactionsItems}</div>
          <TransactionsSum />
        </div>
      </React.Fragment>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    transactions: state.transactions,
    rate: state.rate,
  };
};

const mapDispatchToProprs = {
  removeTransaction: removeTransaction,
};

export default connect(mapStateToProps, mapDispatchToProprs)(TransactionsAll);
