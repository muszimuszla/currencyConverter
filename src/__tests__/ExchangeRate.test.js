import React from "react";
import { shallow } from "enzyme";
import { ExchangeRate } from "../components/ExchangeRate";

let enzymeWrapper;
const changeExchangeRate = jest.fn();

function setup() {
  const props = {
    changeExchangeRate,
    rate: 4.4,
  };

  enzymeWrapper = shallow(<ExchangeRate {...props} />);

  return {
    props,
    enzymeWrapper,
  };
}

describe("Exchange Rate", () => {
  it("renders without crashing", () => {
    setup();
  });

  it("renders correctly", () => {
    expect(enzymeWrapper).toMatchSnapshot();
  });

  it("input renders correct exchange rate value", () => {
    const tmp = setup();
    expect(enzymeWrapper.find('input[type="number"]').props().value).toBe(
      tmp.props.rate
    );
  });

  it("input change rendered value", () => {
    const mockEvent = { target: { value: 12.3 } };
    enzymeWrapper.find('input[type="number"]').simulate("change", mockEvent);
    expect(changeExchangeRate).toHaveBeenCalledWith(12.3);
    changeExchangeRate.mockClear();
  });

  it("input accepts only digits", () => {
    const mockEvent = { target: { value: "abc" } };
    enzymeWrapper.find('input[type="number"]').simulate("change", mockEvent);
    expect(changeExchangeRate).toHaveBeenCalledTimes(0);
    changeExchangeRate.mockClear();
  });
});
