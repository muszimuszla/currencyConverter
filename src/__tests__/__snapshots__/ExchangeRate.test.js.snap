// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Exchange Rate renders correctly 1`] = `
<Fragment>
  <div
    className="exchange-rate"
  >
    <p>
      1 EUR = 
    </p>
    <input
      min="0"
      name="rate"
      onChange={[Function]}
      step="0.01"
      type="number"
      value={4.4}
    />
    <p>
      PLN 
    </p>
  </div>
</Fragment>
`;
